server/internal/server/ollama_client.go

```diff
@@ -1,72 +1,133 @@
 package server
 
 import (
+	"bufio"
+	"bytes"
 	"context"
+	"encoding/json"
+	"fmt"
 	"log"
+	"net/http"
+	"net/url"
+	"os"
 
 	v1 "github.com/llm-operator/inference-manager/api/v1"
-	"github.com/ollama/ollama/api"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
-func newClient() (*api.Client, error) {
-	client, err := api.ClientFromEnvironment()
-	if err != nil {
-		return nil, err
+// Modified based on https://github.com/ollama/ollama/blob/main/api/client.go
+// Support OpenAI compatible API, and avoid converting message format conversion.
+
+// client encapsulates client state for interacting with the ollama service.
+type client struct {
+	base *url.URL
+	http *http.Client
+}
+
+func newClient() *client {
+	return &client{
+		base: &url.URL{
+			Scheme: "http",
+			Host:   os.Getenv("OLLAMA_HOST"),
+		},
+		http: http.DefaultClient,
 	}
-	return client, nil
 }
 
-func handleChatRequest(ctx context.Context, req *v1.CreateChatCompletionRequest) (*v1.ChatCompletion, error) {
-	client, err := newClient()
+const maxBufferSize = 512 * 1000
+
+func (c *client) stream(ctx context.Context, method, path string, data any, fn func([]byte) error) error {
+	var buf *bytes.Buffer
+	if data != nil {
+		bts, err := json.Marshal(data)
```

Comment:
I might not be reading the code right, but will this work? I thought we're passing a proto message, not json.

Reply:
seems working

```
kubectl port-forward -n llm-operator services/llm-operator-inference-manager-server-http 8444 &

curl http://localhost:8444/v1/chat/completions -d '{
  "model": "google-gemma-2b-it-q4",
  "messages": [{"role": "user", "content": "Why is the sky blue?"}]
}'
{"id":"chatcmpl-246", "choices":[{"finishReason":"stop", "index":0, "message":{"content":"The sky appears blue due to Rayleigh scattering. This scattering of light occurs when light interacts with molecules in the Earth's atmosphere.\n\n**Rayleigh scattering** is a process in which light waves are scattered in different directions by tiny particles in the medium through which they travel.\n\nIn the case of the sky, the molecules that scatter light are molecules of nitrogen and oxygen in the air. These molecules are much smaller than the wavelengths of visible light, which means that they scatter blue light more efficiently than other colors.\n\nWhen sunlight enters the atmosphere, it interacts with these molecules and is scattered in all directions. The blue light is scattered more strongly due to its shorter wavelength, resulting in the blue color that we see in the sky.\n\nOther factors that can influence the perceived color of the sky include:\n\n* **Dust particles:** Dust particles in the atmosphere can also scatter light, which can affect the color of the sky.\n* **Clouds:** Clouds can also scatter light, which can change the color of the sky.\n* **Sun position:** The position of the sun relative to the observer can also influence the color of the sky.\n\nOverall, the sky appears blue due to Rayleigh scattering of light by atmospheric molecules.", "toolCalls":[], "role":"assistant"}, "logprobs":null}], "created":1714586628, "model":"google-gemma-2b-it-q4", "systemFingerprint":"fp_ollama", "object":"chat.completion", "usage":{"completionTokens":248, "promptTokens":15, "totalTokens":263}}
```

Reply:
I see. I guess this works as we also use the same message for gRPC gateway.